msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/NativeAuthPublicClientApplicationAnotherAbstractTest.kt

```diff
@@ -0,0 +1,156 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.client.e2e.tests.network.nativeauth
+
+import android.app.Activity
+import android.content.Context
+import androidx.test.core.app.ApplicationProvider
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+import com.microsoft.identity.client.ILoggerCallback
+import com.microsoft.identity.client.Logger
+import com.microsoft.identity.client.PublicClientApplication
+import com.microsoft.identity.client.e2e.shadows.ShadowAndroidSdkStorageEncryptionManager
+import com.microsoft.identity.client.e2e.tests.IPublicClientApplicationTest
+import com.microsoft.identity.client.exception.MsalException
+import com.microsoft.identity.common.internal.controllers.CommandDispatcherHelper
+import com.microsoft.identity.internal.testutils.TestUtils
+import com.microsoft.identity.internal.testutils.labutils.KeyVaultFetchHelper
+import com.microsoft.identity.internal.testutils.labutils.LabConstants
+import com.microsoft.identity.internal.testutils.labutils.LabUserHelper
+import com.microsoft.identity.internal.testutils.labutils.LabUserQuery
+import com.microsoft.identity.internal.testutils.nativeauth.ConfigType
+import com.microsoft.identity.internal.testutils.nativeauth.api.models.NativeAuthTestConfig
+import com.microsoft.identity.nativeauth.INativeAuthPublicClientApplication
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.test.StandardTestDispatcher
+import kotlinx.coroutines.test.setMain
+import org.junit.After
+import org.junit.Assert
+import org.junit.Before
+import org.junit.runner.RunWith
+import org.mockito.Mockito
+import org.robolectric.RobolectricTestRunner
+import org.robolectric.annotation.Config
+import org.robolectric.annotation.LooperMode
+
+// TODO: move to "PAUSED". A work in RoboTestUtils will be needed though.
+@LooperMode(LooperMode.Mode.LEGACY)
+@RunWith(RobolectricTestRunner::class)
+@Config(shadows = [ShadowAndroidSdkStorageEncryptionManager::class])
+abstract class NativeAuthPublicClientApplicationAnotherAbstractTest : IPublicClientApplicationTest {
```

Comment:
@Yuki-YuXin Could you please clarify why you created a new abstract class for native auth public client application? I can see that a lot of code is repeated between this class and the NativeAuthPublicClientApplicationAbstractTest.kt file

Reply:
The new abstract interface is to include more config and challenge situations based on the sign up test case needs. Meanwhile, there are still sign in files need the original abstract class. They are waiting for the next phase of work.

Reply:
It's not clear to me why we need to create a new class. Could you include "more config and challenge situations" in the existing abstract class?

Reply:
The original structure of abstract class can only have one set of config type and challenge types setting, which does not satisfy the needs of this sign up work. With the new structure of new abstract class, the input of config and challenge type is more flexible.

For example.
```
        config = getConfig(defaultConfigType)
        application = setupPCA(config, defaultChallengeTypes)
```
and 

```
        config = getConfig(ConfigType.SIGN_UP_PASSWORD)
        application = setupPCA(config, listOf("oob"))
```

Reply:
Can you make the input of config and challenge type of "NativeAuthPublicClientApplicationAbstractTest" more flexible? In this way we don't have to create a clone of this class? I don't think it's a good idea to have two almost identical classes:
- NativeAuthPublicClientApplicationAbstractTest
- NativeAuthPublicClientApplicationAnotherAbstractTest

Reply:
NativeAuthPublicClientApplicationAbstractTest is singleton pattern, which only accept one config. The improvement of NativeAuthPublicClientApplicationAnotherAbstractTest is to make it more flexible on config setting. The reason we have two almost identical classes is because the remaining classes except than sign up like sign in are still applied the singleton one. At last, there would be only one abstract class once we finished all the automation work.

Reply:
@Yuki-YuXin if refactoring the original `NativeAuthPublicClientApplicationAbstractTest` is too much work for this PR, can you add a comment saying that the work will be done as part of a different PR, and create a PBI for that? Let's try to target that PBI for this quarter, just right after this PR gets merged.

Reply:
Could all tests (including the existing ones) not use the initialiser in `NativeAuthPublicClientApplicationAbstractTest`? If so, can't we delete the old one and make all tests use the new one? Or somehow prevent the duplicated code by subclassing?
I see these 2:
```
    private val defaultConfigType = ConfigType.SIGN_UP_OTP
    private val defaultChallengeTypes = listOf("password", "oob")
```
Being used in a lot of tests, and if that's why the class has to be duplicated, can't they be moved into a config file or made global so all tests can use them?

Reply:
Hi @spetrescu84 , the first batch of E2E tests were all happy scenarios (no errors in the process), whereas this batch of E2E tests had many unhappy scenarios due to misconfiguration. The correct configuration of the first batch of E2E tests are all initialized in the` NativeAuthPublicClientApplicationAbstractTest`.  After add the second batch, `SignUpEmailOTPTest` has tests using the correct configuration as well as tests using the wrong configuration (And they're wrong about different things). So there's no way to globalize the configuration, only to move it out of the ` NativeAuthPublicClientApplicationAbstractTest`

Reply:
@nilo-ms @spetrescu84 I'm resolving this conversation as Yuki has merged everything into the same NativeAuthPublicClientApplicationAbstractTest.

Feel free to unresolve it if you think there's something else to do.